akka-actor/src/main/scala/akka/dispatch/affinity/AffinityPool.scala

```diff
@@ -6,28 +6,114 @@ package akka.dispatch.affinity
+import akka.event.Logging
+import akka.util.{ ImmutableIntMap, OptionVal, ReentrantGuard }
 
+import scala.annotation.{ tailrec, switch }
 import scala.collection.mutable
 import scala.util.control.NonFatal
 
+@InternalApi
+@ApiMayChange
+private[affinity] object AffinityPool {
+  type PoolState = Int
+  // PoolState: waiting to be initialized
+  final val Uninitialized = 0
+  // PoolState: currently in the process of initializing
+  final val Initializing = 1
+  // PoolState: accepts new tasks and processes tasks that are enqueued
+  final val Running = 2
+  // PoolState: does not accept new tasks, processes tasks that are in the queue
+  final val ShuttingDown = 3
+  // PoolState: does not accept new tasks, does not process tasks in queue
+  final val ShutDown = 4
+  // PoolState: all threads have been stopped, does not process tasks and does not accept new ones
+  final val Terminated = 5
+
+  // Method handle to JDK9+ onSpinWait method
+  private val onSpinWaitMethodHandle =
+    try
+      OptionVal.Some(MethodHandles.lookup.findStatic(classOf[Thread], "onSpinWait", methodType(classOf[Unit])))
+    catch {
+      case NonFatal(_) ⇒ OptionVal.None
+    }
+
+  type IdleState = Int
+  // IdleState: Initial state
+  final val Initial = 0
+  // IdleState: Spinning
+  final val Spinning = 1
+  // IdleState: Yielding
+  final val Yielding = 2
+  // IdleState: Parking
+  final val Parking = 3
+
+  // Following are auxiliary class and trait definitions
+  private final class IdleStrategy(idleCpuLevel: Int) {
+
+    private[this] val maxSpins = 1100 * idleCpuLevel - 1000
+    private[this] val maxYields = 5 * idleCpuLevel
+    private[this] val minParkPeriodNs = 1
+    private[this] val maxParkPeriodNs = MICROSECONDS.toNanos(250 - ((80 * (idleCpuLevel - 1)) / 3))
+
+    private[this] var state: IdleState = Initial
+    private[this] var turns = 0L
+    private[this] var parkPeriodNs = 0L
+    @volatile private[this] var idling = false
+
+    @inline private[this] final def transitionTo(newState: IdleState): Unit = {
+      state = newState
+      turns = 0
+    }
+
+    final def isIdling: Boolean = idling
+
+    final def idle(): Unit = {
+      (state: @switch) match {
+        case Initial ⇒
+          idling = true
+          transitionTo(Spinning)
+        case Spinning ⇒
+          onSpinWaitMethodHandle match {
+            case OptionVal.Some(m) ⇒ m.invokeExact()
+            case OptionVal.None    ⇒
+          }
+          turns += 1
+          if (turns > maxSpins)
+            transitionTo(Yielding)
+        case Yielding ⇒
+          turns += 1
+          if (turns > maxYields) {
+            parkPeriodNs = minParkPeriodNs
+            transitionTo(Parking)
+          } else Thread.`yield`()
+        case Parking ⇒
+          LockSupport.parkNanos(parkPeriodNs)
+          parkPeriodNs = Math.min(parkPeriodNs << 1, maxParkPeriodNs)
+      }
+    }
+
+    final def reset(): Unit = {
+      idling = false
```

Comment:
volatile write here is actually beneficial because the purpose of the method is to stall things a bit ;)